### ISSUE TITLE
Submission Display Options for Notebook/Upload Gradeables

### ISSUE THREAD

**Issue by bmcutler** [Submitty/Submitty#5345]
The submission/view grade page for students has several visual objects that should be optionally displayed depending on how various gradeable options are configured.

We should discuss the objects of this page, document how they are controlled by instructor configuration settings, give intuitive control to the instructor for these objects.

1. due date & time (with time zone) 
   * should only be displayed if the assignment has a due date

2. [instructor selection: normal submission / make submission for student / bulk upload ]
   * only visible for instructor users
   * current placement is an afterthought, probably should revise
   * should change this to:
       * a selector [ normal submission / make submission for student ] for normal uploads, or
       * no selector (always bulk upload) for bulk uploads

3. draw & drop zone
   * (correctly) does not appear for vcs assignments
   * this should not be visible for notebook gradeables that don't require a 
      supplemental file upload
      (the student submission will only be multiple choice / short answer)
       * there is currently (?) no way for instructors to control appearance of this box 
          (that is, to specify if their assignment requires supplemental file upload)
   * this option should not appear for students for bulk upload

4. submission version / active version drop down and cancel/do not grade options
   * this should not appear for students for bulk upload

5. submitted files for the selected version w/ download option
  * this should not appear for notebook gradeables with no supplemental file upload
  * if it is a notebook with supplemental file upload, it should filter out the files created 
     for multiple choice/short answer & just show the supplemental files

6. submission timestamp, autograding time, gradeable access duration
   * don't show any of this for bulk upload gradeable
   * don't show autograding time when there are no autograding test cases (#3683)
   * don't show access duration if this is not a timed notebook gradeable

7. autograding
   * don't show autograding time when there are no autograding test cases

8. hidden autograding tests..
   * don't show when grades are not yet released

9. ta grading
   * don't show when there is no ta grading, or when (ta) grades are not yet released

10. peer grading
   * don't show when there is no peer grading, or when grades are not yet released

----------

these options in the autograding configuration are unintuitive.  Some of these should be set already by selecting "bulk upload", not be embedded in the autograding config (which can be misused on regular hw and cause confusion)

https://github.com/Submitty/Submitty/blob/master/more_autograding_examples/pdf_exam/config/config.json

    "hide_submitted_files" : true,
    "hide_version_and_test_details" : true

also in:
https://github.com/Submitty/Submitty/blob/master/more_autograding_examples/notebook_basic/config/config.json

----------

This list & discussion is incomplete
@emaicus & @Kevin-MM & others let's  discuss & revise

**Comment by bmcutler** [comment]
related to #4374

**Comment by Kevin-MM** [comment]
I'd like to introduce the concept of a 'widget' or 'display object' to submitty.   I think it would be great to more thoroughly model anything that could be displayed (and perhaps displayed more than once per submission page).  This could include multiple choice inputs, drop zones, submit buttons, you name it.  @shailpatels has been doing a great job moving some of the notebook items into models.  I'm unsure how well modeled other things are like the drop zones.  Some things currently to the best of my knowledge aren't really modeled at like like the submit buttons.

There also seems to be a bit of a gray area here about what is a notebook object and what is not.  For example I don't think drop zones were originally part of the notebook spec, but were ultimately needed and kind of made their way in there in a non-consistent way.  We will need to more clearly define the roles of each specific display object.  IE, we need to clearly decide what a drop zone is, is it a notebook cell, is it something else?  For this proposal I think we need to be thinking at a level above the notebooks, since we will need to render objects that aren't necessarily part of the notebook.

Once modeling is complete I propose they implement a `DisplayObject` interface.  This interface would probably only have one method, maybe called `compile` or `execute` or `view` or something similar.  This method would return a block of html code, which we then append to the display buffer.  I'm not sure of Twigs capabilities but I imagine it should be possible to create a twig file for each individual class implementing the `DisplayObject` interface.

So to recap, we could have a flow that goes something like this
1. `SubmissionController` parses the configuration data to determine which information needs to be displayed, and constructs an array of objects that implement `DisplayObject`.  Any configurable options for the object would happen at this step, probably just by passing them as parameters to the class constructor.
1. The view file receives this array and goes over it in a loop, calling the `view()` method.  The result of `view()` call is HTML code, which gets appended to a buffer we build.
1. The buffer we built at this stage is appended to the output buffer being build by `Output.php`

**Comment by bmcutler** [comment]
@Kevin-MM 

I'll toss a feature request into the design discussion.  Some quiz/exam formats could benefit from  drop zones interspersed with multiple choice/short answer questions.  I wrote a quiz with 6 problems, mostly short answer, some problems had multiple parts.  For one of the problems, one part was to draw a diagram (e.g., draw it on paper & take a photo/scan and upload the file).  It would be nice in that case to position the drop zone with the problem, in the middle of other notebook cells.  

Sort of related is the idea of multiple submit buttons throughout the notebook, encouraging students to submit/save incrementally as they progress through the exam.  Just to make sure they don't lose their work, but also in some cases to get feedback if autograding results are available and configured to be displayed before the due date.
